### PR TITLE
Project 2: Register Allocation for Bril

### PR DESCRIPTION
**Pull request by xu3kev**
This is the blog post for #68.

### CODE REVIEW

File: content/blog/2019-10-23-register-allocation/index.md [hunk 1]
```diff
@@ -0,0 +1,142 @@
++++
+title = "Register Allocation for Bril"
+extra.author = "Hongbo Zhang, Sachille Atapattu, Wen-Ding Li"
+extra.bio = """
+  Hongbo Zhang is a first PhD student in computer science. He is interested in systems and computer architectures. He is also okay archer shooting recurve bow.
+  [Wen-Ding Li](https://www.cs.cornell.edu/~wdli/) is a Ph.D. student at Cornell interested in high-performance code generation.
```
Comment (sampsyo) [inline content/blog/2019-10-23-register-allocation/index.md]: Missing bio for Sachille.

File: content/blog/2019-10-23-register-allocation/index.md [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+"""
++++
+# Register Allocation for Bril
+
+Processors usually have registers which can be used to store variables for faster access. Program will run faster if we can put variables in these registers. Nevertheless, registers are limited, so often times we still need to access memory (or cache) instead of registers and memory access are costly compared with register access. Although modern program often use large number of variables, we can analyze their live range so the same registers may be able to hold different varialbes throughout the program. Then, we can put more varialbes in the registers which may lead to fewer memory access and get a faster program.
```
Comment (sampsyo) [inline content/blog/2019-10-23-register-allocation/index.md]: Program will run faster -> programs
often times -> oftentimes
memory access are costly -> accesses
modern program often use -> programs
their live range -> ranges
varialbes -> variables (twice)
fewer memory access -> accesses
get a faster program -> *yield* a faster program

Throughout your post, please do these things before requesting another review:
- Run spell check.
- Be careful about plural vs. singular nouns and pronouns.

File: content/blog/2019-10-23-register-allocation/index.md [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+  [Wen-Ding Li](https://www.cs.cornell.edu/~wdli/) is a Ph.D. student at Cornell interested in high-performance code generation.
+"""
++++
+# Register Allocation for Bril
+
+Processors usually have registers which can be used to store variables for faster access. Program will run faster if we can put variables in these registers. Nevertheless, registers are limited, so often times we still need to access memory (or cache) instead of registers and memory access are costly compared with register access. Although modern program often use large number of variables, we can analyze their live range so the same registers may be able to hold different varialbes throughout the program. Then, we can put more varialbes in the registers which may lead to fewer memory access and get a faster program. 
+
+The goal of our project is to perform register allocation on Bril program and anaylze its performance. In this project, we implemented register allocation for Bril via graph coloring. Our project can be found in [this repository](https://github.com/xu3kev/bril) under `regalloc` directory.
```
Comment (sampsyo) [inline content/blog/2019-10-23-register-allocation/index.md]: You can link directly to the directory.

File: content/blog/2019-10-23-register-allocation/index.md [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+
+## Method
+Our approach for the register allocation can be divided in three main parts. First, we perform liveness analysis and then build a graph. Second, we solve approximate k graph coloring problem to obtain a mapping from each varialbe to corresponding register. At last, we generate Bril code based on the mapping.
```
Comment (sampsyo) [inline content/blog/2019-10-23-register-allocation/index.md]: It's usually written as "graph $k$-coloring," not "k graph coloring."

File: content/blog/2019-10-23-register-allocation/index.md [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+
+### Liveness Analysis
+We use data flow analysis to get the live varialbes at each instruction of the program. First we transform the input Bril program to control flow graph (cfg) and then perform liveness analysis with a standard backward flow anaylsis. Once we obtain the live varialbes in the end of each code block, we start from that and go backward: when we see a definition of a varialbe, we kill the variable, and when we see the use of the varialbe, we mark it as alive. Then, we can get the live varialbes at every instructions.
```
Comment (sampsyo) [inline content/blog/2019-10-23-register-allocation/index.md]: Capitalize "CFG" (throughout).

File: content/blog/2019-10-23-register-allocation/index.md [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+
+### Graph Coloring
+With the liveness ranges of all local variables, we can know which variables are "alive" at the same time. If two variables have overlapped liveness ranges, it means that they cannot be allocated to the same register. Otherwise, they may be allocated to the same register. Our goal is to have register assignment so that we can allocate as many variables on registers as possible.
+We solve this question by graph coloring. Each node in the graph represents a local variable in the function. There is an edge between two nodes if the two variables have overlapped liveness range. We just needs to compute a color assignment with at most *k* colors, such that every pair of nodes on the two sides of an edge have different colors.
```
Comment (sampsyo) [inline content/blog/2019-10-23-register-allocation/index.md]: Some more example of plurality corrections:

variable -> variables
range -> ranges
needs -> need

File: content/blog/2019-10-23-register-allocation/index.md [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+Clearly, it is not always possible to find a coloring scheme. For example, it is not possible to allocate 2 registers to 3 variables with overlapped liveness ranges. Maybe we should find a coloring scheme to cover the subgraph with most number of nodes (variables)?
+Unfortunately, it is still impossible. *k*-coloring is a NP-complete problem. Instead of finding the optimal coloring, we use optimistic coloring to find one large subgraph with as many nodes as possible.
```
Comment (sampsyo) [inline content/blog/2019-10-23-register-allocation/index.md]: By "impossible," do you mean "intractable?"

Comment (sampsyo) [inline content/blog/2019-10-23-register-allocation/index.md]: > Instead of finding the optimal coloring, we use optimistic coloring to find one large subgraph with as many nodes as possible.

It's not clear how this gets around the NP-completeness… it seems to say you're solving a *harder* problem, which is finding the largest subgraph that is $k$-colorable.

File: content/blog/2019-10-23-register-allocation/index.md [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+On each interation, we find the node with the least number of neighbours that are not assigned a color yet. Then we try to assign a color to the selected node based on the constraints of all neighbour nodes. If it is not possible to assign a color to the selected node, then we remove the node from the graph along with all edges connected to the node.
+Only the variables with a assigned color will be allocated to the corresponding registers, all reads or writes to other variables will involve a memory load and store.
```
Comment (sampsyo) [inline content/blog/2019-10-23-register-allocation/index.md]: This sounds like it's describing a simple greedy heuristic, right? It might be nice to discuss here whether this heuristic has been analyzed theoretically. Does it have an approximation bound w/r/t the optimal coloring?

File: content/blog/2019-10-23-register-allocation/index.md [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+| alloc1 | Tests a simple 4 variable program taken from [these](https://web.stanford.edu/class/archive/cs/cs143/cs143.1128/lectures/17/Slides17.pdf) slides.    |
+| alloc2 | Test with 14 variables possible to allocate to 4 registers, inspired from the [book](https://www.google.com/books/edition/Modern_Compiler_Implementation_in_Java/N-sgAwAAQBAJ?hl=en&gbpv=0) |
+| alloc3 | Test with 14 variables possible to allocate to 6 registers, adopted from the [book](https://www.google.com/books/edition/Modern_Compiler_Implementation_in_Java/N-sgAwAAQBAJ?hl=en&gbpv=0) |
+| matmul | An implementation of matrix multiplication in Bril                                |
+| dotprod | An implementation of dot product in Bril                                |
+| fib | An implementation of fibonnacci in Bril                                |
+| fact | An implementation of factorial in Bril                                |
+| polymul | An implementation of polynomial multiplication in Bril                    |
+
+The last 5 benchmarks are from this [repository](https://github.com/xu3kev/bril/tree/master/benchmark).
```
Comment (sampsyo) [inline content/blog/2019-10-23-register-allocation/index.md]: As with your previous project, Wen-Ding, nice work concocting some realistic-ish benchmarks!!

File: content/blog/2019-10-23-register-allocation/index.md [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+
+### Baseline
+In order to evaluate our register allocation, we assume that there is no register holding data across different instruction as the baseline, so that each instruction loads all operands from the memory and stores the result back to memory after execution. This is a conservative approach and provides an upper bound to the number of loads and stores that can occur.
```
Comment (sampsyo) [inline content/blog/2019-10-23-register-allocation/index.md]: > we assume that there is no register holding data across different instruction as the baseline

This is a little hard to parse—are you saying that, in your assumed machine model, operations can happen "directly" on memory operands without first going into a register? FWIW, there is a term "register machine" for an ISA where everything has to happen on register operands. What you're saying is essentially the opposite of that, which you might call a [register memory architecture](https://en.wikipedia.org/wiki/Register_memory_architecture).

File: content/blog/2019-10-23-register-allocation/index.md [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+
+We also implemented a naïve register allocation technique, which assigns registers in a round-robin manner. This technique doesn't guarantee any efficiencies, but merely provides a mechanism to map to registers. This implementation presently does not support branch operations and therefore is only used for a subset of benchmarks as a sanity check.
```
Comment (sampsyo) [inline content/blog/2019-10-23-register-allocation/index.md]: > guarantee any efficiencies

Does your graph-coloring algorithm provide any guarantees, in contrast?

Comment (sampsyo) [inline content/blog/2019-10-23-register-allocation/index.md]: Even if the naive version doesn't support branches, wouldn't it be possible to do it one block at a time and just store everything back to memory at the end of the block?

File: content/blog/2019-10-23-register-allocation/index.md [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+
+To gain more insight, we also manually allocated registers by graph coloring for some tests. This provides a reasonable ceiling to potential performance.
```
Comment (sampsyo) [inline content/blog/2019-10-23-register-allocation/index.md]: What does it mean to "manually" use graph coloring? Did you run the exponential-time algorithm in your head? How do you know if you found the optimal solution?

File: content/blog/2019-10-23-register-allocation/index.md [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+
+As for accessing register values (the case where special variables represented registers will not be counted as a memory access):
+
+- `r1: int = const 2;` has 0 memory access operations
+- `a: int = id r2;` has 1 memory access operations (store a)
+- `c:int = add a r2;` has 2 memory access operations (load a, store c)
+
+We compare the total number of memory access operations for the same program with and without register allocation. 
+
+It was identified during evaluation that this implementation of counting doesn't support the `ret` instruction. We hope to fix this issue shortly. We have tested this interpreter version with some hand written register allocated code.
```
Comment (sampsyo) [inline content/blog/2019-10-23-register-allocation/index.md]: Hmm, but wouldn't `ret` just count as zero memory accesses?

File: content/blog/2019-10-23-register-allocation/index.md [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+
+### Simulated load and store counts
+
+We simulated the baseline and graph coloring implementation using Brili to test for two things,
+1. To make sure that the functionality hasn't been affected by our optimization
+2. To measure memory access operations as a proxy for performance
+
+The following tables provides the memory acccess count in the form of `loads/stores`. These tests are run assuming for 4 registers.
```
Comment (sampsyo) [inline content/blog/2019-10-23-register-allocation/index.md]: "assuming for 4 registers" (extra "for")?

File: content/blog/2019-10-23-register-allocation/index.md [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+
+Basline and graph coloring tests are an average from multiple simulation runs. This turned out to be useful for the graph coloring technique, as the register assignment can change.
```
Comment (sampsyo) [inline content/blog/2019-10-23-register-allocation/index.md]: Why can the assignment change? Is some sort of iteration randomized or otherwise nondeterministic? If so, what is the source of the nondeterminism?

File: content/blog/2019-10-23-register-allocation/index.md [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+
+This first table evaluates several handcrafted tests to test whether the grach coloring implementation can handle all instructions and where does it lie in terms of performance in a space from inefficient baseline to optimized manual handtuning. As described in the implementation section, the implementation does not achieve optimal coloring, but manages to reduce memory accesses compared to other techniques.
+
+| Benchmark| baseline  | naïve     | graph coloring |
+|---------|-----------|-----------|----------------|
+| loop    | 53/24     |           | 21/2           |
+| matmul  | 2048/1216 | 1280/1980 | 1008/126       |
+| dotprod | 257/145   | 128/205   | 169/78         |
+| fib     | 103/56    |           | 32/12          |
+| fact    | 33/14     |           | 3/0.3           |
```
Comment (sampsyo) [inline content/blog/2019-10-23-register-allocation/index.md]: Is it possible to have 0.3 of a store?

File: content/blog/2019-10-23-register-allocation/index.md [hunk 17]
```diff
@@ -0,0 +1,142 @@
...
+| polymul | 929/509   | 711/767   | 869/505        |
+
+These benchmark tests stress test the implementations and provide interesting results. As seen from the tables, graph coloring technique manages to consistently have comparable or less accesses than the other two techniques. It also seem to compromise some load operations to avoid store operations.
```
Comment (sampsyo) [inline content/blog/2019-10-23-register-allocation/index.md]:
```suggestion
These benchmark tests stress test the implementations and provide interesting results. As seen from the tables, the graph coloring technique manages to consistently have comparable or less accesses than the other two techniques. It also seems to compromise some load operations to avoid store operations.
```

File: content/blog/2019-10-23-register-allocation/index.md [hunk 18]
```diff
@@ -0,0 +1,142 @@
...
+
+Figure illustrates how effective graph coloring is by percentage reduction in memory accesses. Apart from polymul it shows a significant reduction in loads and stores. A possible reason for poor performance in polymul might be it's high register usage. We intend to analyze this further by analyzing how the percentage reduction in accesses increase with more registers.
```
Comment (sampsyo) [inline content/blog/2019-10-23-register-allocation/index.md]: "Figure illustrates" -> *This* figure illustrates, or *The* figure illustrates, or *The following* figure illustrates, or *Figure 1* illustrates…

Comment (sampsyo) [inline content/blog/2019-10-23-register-allocation/index.md]: it's -> its

**Review by sampsyo — CHANGES_REQUESTED** [review]
Awesome work! This is a neatly done project and a very intelligible writeup. I have a few clarification questions within, along with a proofreading request about plurality. I know plural nouns and verb conjugation are a difficult part of English writing, but with some focused attention, it won't be hard to find and fix these.

File: content/blog/2019-10-23-register-allocation/index.md (hunk 16 above)
Comment (sa2257) [reply content/blog/2019-10-23-register-allocation/index.md]: It's averaged over several runs.

Comment (sampsyo) [reply content/blog/2019-10-23-register-allocation/index.md]: That’d do it!
